src/gui/uithememanager.cpp

```diff
@@ -58,6 +61,98 @@ namespace
 
         return {};
     }
+
+    QByteArray readFile(const QString &fileName)
+    {
+        QFile file(fileName);
+        if (!file.open(QIODevice::ReadOnly | QIODevice::Text))
+        {
+            LogMsg(UIThemeManager::tr("UITheme - Failed to open \"%1\". Reason: %2")
+                    .arg(QFileInfo(fileName).fileName(), file.errorString())
+                   , Log::WARNING);
+            return {};
+        }
+
+        return file.readAll();
+    }
+
+    class QRCThemeSource : public UIThemeSource
+    {
+    public:
+        QByteArray styleSheet() override
+        {
+            return readFile(QRC_THEME_DIR + STYLESHEET_FILE_NAME);
+        }
+
+        QByteArray config() override
+        {
+            return readFile(QRC_THEME_DIR + CONFIG_FILE_NAME);
+        }
+
+        QString iconPath(const QString &iconId) override
+        {
+            return findIcon(iconId, QRC_ICONS_DIR);
+        }
+
+    private:
+        /* static */ const QString QRC_THEME_DIR {":/uitheme/"};
+        /* static */ const QString QRC_ICONS_DIR = QRC_THEME_DIR + THEME_ICONS_DIR;
+    };
+
+    class FolderThemeSource : public UIThemeSource
+    {
+    public:
+        FolderThemeSource(const QString &path)
+            : m_folder {path}
+            , m_iconsDir {m_folder.absolutePath() + "/icons/"}
+        {
+        }
+
+        QByteArray styleSheet() override
+        {
+            QByteArray styleSheetData = readFile(m_folder.absoluteFilePath(STYLESHEET_FILE_NAME));
+            return styleSheetData.replace(STYLESHEET_ICONS_DIR, (m_folder.absolutePath() + '/').toUtf8());
+        }
+
+        QByteArray config() override
+        {
+            return readFile(m_folder.absoluteFilePath(CONFIG_FILE_NAME));
+        }
+
+        QString iconPath(const QString &iconId) override
+        {
+            return findIcon(iconId, m_iconsDir);
+        }
+
+    private:
+        const QDir m_folder;
+        const QString m_iconsDir;
+    };
+
+
+    std::unique_ptr<UIThemeSource> createUIThemeSource(QObject *context)
+    {
+        const Preferences *const pref = Preferences::instance();
+        if (!pref->useCustomUITheme())
```

Comment:
You shouldn't check this setting in two places of `UIThemeManager` initialization.
And this is not some kind of general-purpose function, so that you can argue that the specific context of its use is unknown here. Just create `UIThemeSource` here (in case it is called) and let the `UIThemeManager` to decide whether it should be created or not.